x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -350,53 +308,6 @@ public void prepareIndexIfNeededThenExecute(final Consumer<Exception> consumer,
             } else if (indexState.indexExists() && indexState.isIndexUpToDate == false) {
```

Comment:
The changes around here are neat given that this is a refactoring! 👍 
But I think we should let the runnable through if the index does not exist, and only then.
If the index exists, and it has an old mapping version we should hold off the runnable, because it can race with the service that updates the mapping. Related side note: can the mapping update service maybe hook into the "auto put mapping action" in a similar fashion that the system index creation hooks into the "index auto create action"?

Reply:
Not to dwell to much on it, but IMO I think it is worth investing in adapting the auto mapping update action. It is added recently, so I don't expect there to be too obscure behaviors encoded. The reason I mention it is that moving the mapping update from a ClusterStateListener to another (from `SecurityIndexManager` to `SystemIndexManager`) is not a great step forward.

Just a suggestion, I haven't investigated it carefully, and it's outside security's concern.